File: src/test/java/info/kfgodel/function/MemoizedSupplierTest.java
package info.kfgodel.function;

import ar.com.dgarcia.javaspec.api.JavaSpec;
import ar.com.dgarcia.javaspec.api.JavaSpecRunner;
import org.junit.runner.RunWith;

import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;


/**
 * This class verifies that the memorize supplier behaves as expected
 * Date: 04/05/19 - 18:30
 */
@RunWith(JavaSpecRunner.class)
public class MemoizedSupplierTest extends JavaSpec<FunctionTestContext> {
  @Override
  public void define() {
    describe("a memoized supplier", () -> {
      test().memoized(() -> MemoizedSupplier.create(() -> new AtomicInteger().incrementAndGet()));

      it("returns the first supplied value when called", () -> {
        Object firstValue = test().memoized().get();
        assertThat(firstValue).isEqualTo(1);
      });

      it("keeps returning the first value after the first call",()->{
        Object firstValue = test().memoized().get();
        Object secondValue = test().memoized().get();
        Object thirdValue = test().memoized().get();
        assertThat(secondValue).isEqualTo(1);
        assertThat(thirdValue).isEqualTo(1);
      });
    });

  }
}

File: src/test/kotlin/info/kfgodel/function/MemoizedSupplierTest.kt
package info.kfgodel.dyna

import info.kfgodel.function.MemoizedSupplier
import info.kfgodel.jspek.api.JavaSpecRunner
import info.kfgodel.jspek.api.KotlinSpec
import org.assertj.core.api.Assertions.assertThat
import org.junit.runner.RunWith
import java.util.concurrent.atomic.AtomicInteger

/**
 * Date: 28/3/21 - 02:00
 */
@RunWith(JavaSpecRunner::class)
class ReadmeExampleKotlinTest : KotlinSpec() {
  override fun define() {
    describe("a memoized supplier") {
      val memoized by let { MemoizedSupplier.create { AtomicInteger().incrementAndGet() } }

      it("returns the first supplied value when called") {
        val firstValue: Any = memoized().get()
        assertThat(firstValue).isEqualTo(1)
      }
      it("keeps returning the first value after the first call") {
        val firstValue: Any = memoized().get()
        val secondValue: Any = memoized().get()
        val thirdValue: Any = memoized().get()

        assertThat(secondValue).isSameAs(firstValue)
        assertThat(thirdValue).isSameAs(firstValue)
      }
    }
  }
}
